controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +59,780 @@ ControllerManager::load_controller(
     [&](auto loader)
     {return loader->is_available(controller_type);});
 
-  std::shared_ptr<controller_interface::ControllerInterface> controller(nullptr);
+  controller_interface::ControllerInterfaceSharedPtr controller(nullptr);
   if (it != loaders_.cend()) {
     controller = (*it)->create(controller_type);
   } else {
     const std::string error_msg("Loader for controller '" + controller_name + "' not found\n");
     RCLCPP_ERROR(get_logger(), "%s", error_msg.c_str());
     throw std::runtime_error(error_msg);
   }
+  ControllerSpec controller_spec;
+  controller_spec.c = controller;
+  controller_spec.info.name = controller_name;
+  controller_spec.info.type = controller_type;
+
+  return add_controller_impl(controller_spec);
+}
+
+controller_interface::ControllerInterfaceSharedPtr ControllerManager::load_controller(
+  const std::string & controller_name)
+{
+  const std::string param_name = controller_name + ".type";
+  std::string controller_type;
+  if (!has_parameter(param_name)) {
+    declare_parameter(param_name, rclcpp::ParameterValue());
```

Comment:
This is a combination of bad naming of `has_parameter` and my lack of comments here.

`has_parameter` returns whether a parameter is declared or not... Yeah I know...

I'm adding comments to explain the whole parameter thing, but the idea is based of costmap's implementation.
Since a priori we don't know the name of the controllers that will be loaded, we cannot declare them, so when we're told to load a controller by name, we need to declare the parameter if we haven't done so, and then read it.

Reply:
A user via a launch file, or a configuration file. 

Similar to this in ROS1: https://github.com/pal-robotics/tiago_robot/blob/kinetic-devel/tiago_controller_configuration/config/joint_trajectory_controllers.yaml

Or this in ROS2 for costmap layers: https://github.com/ROBOTIS-GIT/turtlebot3/blob/foxy-devel/turtlebot3_navigation2/param/waffle.yaml#L93

Reply:
See comment above. I think the heuristic is fair enough, to load the type of a controller from a node which is called equivalently to the controller. But we must not declare this parameter on the controller manager node itself.

Reply:
I replied above, please take a look.